### PR TITLE
Add support for Array subqueries

### PR DESCRIPTION
**Pull request by jkosh44**
Postgres allows array construction through a query, called Array
subqueries. The syntax is `ARRAY(<query>)`. The result is an array
where each element is a row from the query provided. This commit
implements this feature. Currently it has the limitation that the
result of the query itself can't be an array. The reason is that array
subqueries use the array_agg function, which doesn't support nested
arrays.

This commit was modeled heavily off of #8571, which implements
the same feature for Lists.

### Motivation

This PR adds a feature that has not yet been specified, see above. 

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

**Comment by jkosh44** [comment]
Functionality for this is done, I'm just trying to see where I can eliminate duplication between this and List subqueries. I gave up for now on supporting nested Arrays and instead opened this issue: https://github.com/MaterializeInc/materialize/issues/10701

**Review by sploiselle — COMMENTED** [review]
Talked to Joe about this and he's well underway to generalizing this implementation w/ the LIST subquery code, so going to hold off on really digging into this until he has more of that work done.

**Comment by jkosh44** [comment]
@sploiselle Should be ready for review now, I've de-duplicated the code between array subquery and list subquery planning. Not super happy about the names I came up with so feel free to suggest better ones.

File: src/sql-parser/tests/testdata/select
Comment (sploiselle) [inline src/sql-parser/tests/testdata/select:1233]: We should also include that the CTE can be referenced; is this from a parallel `list` test?

File: test/sqllogictest/array_subquery.slt
Comment (sploiselle) [inline test/sqllogictest/array_subquery.slt:146]: ditto comment about reference; maybe that doesn't matter, as much, or is intentionally testing this behavior?

**Review by sploiselle — APPROVED** [review]
LGTM modulo comments about tests, which I'd like to see fixed

File: test/sqllogictest/array_subquery.slt
Comment (jkosh44) [reply test/sqllogictest/array_subquery.slt]: Agreed, I moved this to `arrays.slt`

File: test/sqllogictest/array_subquery.slt
Comment (jkosh44) [reply test/sqllogictest/array_subquery.slt]: Changed to query

File: src/sql-parser/tests/testdata/select
Comment (jkosh44) [reply src/sql-parser/tests/testdata/select:1243]: I added a second test that references the CTE

File: src/sql/src/plan/query.rs
Comment (jkosh44) [reply src/sql/src/plan/query.rs]: Just added details to the error

**Pull request merged**
